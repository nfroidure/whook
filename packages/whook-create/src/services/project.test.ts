/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, it, beforeEach, jest, expect } from '@jest/globals';
import _inquirer from 'inquirer';
import initProject from './project.js';
import { YError } from 'yerror';
import { type LogService, type LockService } from 'common-services';

describe('initProject', () => {
  const CWD = '/home/whoiam/projects/';
  const inquirer = { prompt: jest.fn<any>() };
  const lock = {
    take: jest.fn<LockService<unknown>['take']>(),
    release: jest.fn<LockService<unknown>['release']>(),
  };
  const ensureDir = jest.fn<any>();
  const log = jest.fn<LogService>();

  beforeEach(() => {
    lock.take.mockReset();
    inquirer.prompt.mockReset();
    lock.release.mockReset();
    ensureDir.mockReset();
    log.mockReset();
  });

  it('should work', async () => {
    lock.take.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockResolvedValueOnce({
      projectName: 'super-project',
    });
    inquirer.prompt.mockResolvedValueOnce({
      projectDirectory: '/home/whoiam/projects/yolo',
    });
    lock.release.mockResolvedValueOnce(undefined);
    ensureDir.mockResolvedValueOnce(undefined);

    const project = await initProject({
      inquirer: inquirer as unknown as typeof _inquirer,
      CWD,
      ensureDir,
      lock,
      log,
    });

    expect({
      project,
      inquirerPromptCalls: inquirer.prompt.mock.calls,
      lockTakeCalls: lock.take.mock.calls,
      lockReleaseCalls: lock.release.mock.calls,
      ensureDirCalls: ensureDir.mock.calls,
      logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
    }).toMatchSnapshot();
  });

  it('should fail with access problems', async () => {
    lock.take.mockResolvedValueOnce(undefined);
    lock.release.mockResolvedValueOnce(undefined);
    inquirer.prompt.mockResolvedValueOnce({
      projectName: 'super-project',
    });
    inquirer.prompt.mockResolvedValueOnce({
      projectDirectory: '/home/whoiam/projects/yolo',
    });
    ensureDir.mockRejectedValueOnce(new YError('E_ACCESS'));

    try {
      await initProject({
        inquirer: inquirer as unknown as typeof _inquirer,
        CWD,
        ensureDir,
        lock,
        log,
      });
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        errorCode: (err as YError).code,
        errorParams: (err as YError).params,
        inquirerPromptCalls: inquirer.prompt.mock.calls,
        lockTakeCalls: lock.take.mock.calls,
        lockReleaseCalls: lock.release.mock.calls,
        ensureDirCalls: ensureDir.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });
});
