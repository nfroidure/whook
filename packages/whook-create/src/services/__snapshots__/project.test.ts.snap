// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initProject should fail with access problems 1`] = `
{
  "ensureDirCalls": [
    [
      "/home/whoiam/projects/yolo",
    ],
  ],
  "errorCode": "E_PROJECT_DIR",
  "errorParams": [
    "/home/whoiam/projects/yolo",
  ],
  "inquirerPromptCalls": [
    [
      [
        {
          "default": "new-whook-project",
          "message": "What's this new project name",
          "name": "projectName",
          "type": "input",
        },
      ],
    ],
    [
      [
        {
          "default": "/home/whoiam/projects/super-project",
          "message": "Provide the project's directory",
          "name": "projectDirectory",
          "type": "input",
        },
      ],
    ],
  ],
  "lockReleaseCalls": [
    [
      "cli:input",
    ],
    [
      "cli:input",
    ],
  ],
  "lockTakeCalls": [
    [
      "cli:input",
    ],
  ],
  "logCalls": [
    [
      "debug",
      "üèÅ - Initializing project...",
    ],
    [
      "error",
      "Cannot create the project's directory: "/home/whoiam/projects/yolo"",
    ],
  ],
}
`;

exports[`initProject should work 1`] = `
{
  "ensureDirCalls": [
    [
      "/home/whoiam/projects/yolo",
    ],
  ],
  "inquirerPromptCalls": [
    [
      [
        {
          "default": "new-whook-project",
          "message": "What's this new project name",
          "name": "projectName",
          "type": "input",
        },
      ],
    ],
    [
      [
        {
          "default": "/home/whoiam/projects/super-project",
          "message": "Provide the project's directory",
          "name": "projectDirectory",
          "type": "input",
        },
      ],
    ],
  ],
  "lockReleaseCalls": [
    [
      "cli:input",
    ],
  ],
  "lockTakeCalls": [
    [
      "cli:input",
    ],
  ],
  "logCalls": [
    [
      "debug",
      "üèÅ - Initializing project...",
    ],
  ],
  "project": {
    "directory": "/home/whoiam/projects/yolo",
    "name": "super-project",
  },
}
`;
